rfcs/0001-task-chunking.md

```diff
@@ -0,0 +1,452 @@
+to support implementing task stickiness for applications with scripting support.
+Many applications support command-line arguments for a contiguous or non-contiguous range
+of frames to process, without the flexible scripting needed for task stickiness.
+To support this diversity of application workload patterns, Open Job Description would benefit
+by implementing both patterns. The combination of the two can be better than either alone,
+when a single task cannot use all available CPU cores, but multiple tasks can run
+in parallel on a host.
+
+This proposal supports two ways to select the size of chunks. The first option is to
+determine all the chunks when submitting the job, based on the required `defaultTaskCount`
+field. If the selected value is too small, the job will run inefficiently because too
+much time is spent on loading compared to running the tasks. If the selected value is too
+large, each chunk takes long to process and cannot be load balanced across a larger fleet
+of worker hosts. Therefore, the proposal provides an optional `targetRuntimeSeconds` field
+specifying how long the user wants each chunk to take. The scheduler can start with
+chunks sized according to `defaultTaskCount`, and then adjust it up or down to be closer
+to the target runtime.
+
+Initial render jobs typically use a contiguous interval of frame numbers like "1-100",
+but then it can be necessary to render an arbitrary set of pick up frames like "7,12-14,95"
+as a follow-up. This proposal makes it straightforward to write job templates that support
+both patterns in simple syntax using a job parameter named `Frames` accepting
+an integer range expression. The `rangeConstraint` field permits the flexibility of
+noncontiguous chunks as well when the application interface supports it.
+
+### Performance and efficiency
+
+The purpose of chunking is to improve the performance and efficiency of jobs running
+on worker hosts. We can model a chunk's runtime to help reason about that. The ratio
+of load time to task runtime is an important input to choosing appropriate
+chunks sizes. This ratio varies widely in practice, but can be quite large in GPU workloads
+such as rendering with a game engine. These diagrams show a ratio large enough to
+make its impact clear visually.
+
+Some runtime data from a test Adobe After Effects rendering workload show a roughly ten times
+performance improvement from 16m51s to 1m41s between invoking `aerender` one frame at a time
+or in chunks of 50 frames. See
+[this pull request description](https://github.com/aws-deadline/deadline-cloud-for-after-effects/pull/77)
+for details.
+
+Our model of chunk runtime starts an application and loads a scene, then runs tasks in parallel
+until all the tasks in the chunk are done.
+
+![chunk example runtime](.images/0001-task-chunking-chunk-example-runtime.svg)
+
+Here's an example timeline for running 15 tasks with this chunking on one worker host,
```

Comment:
Does these 15 tasks mean 15 frames here? or openjd Tasks? I am thinking that we are talking about 3 openjd tasks with 6 frames chunk size but I am a bit confused about the terminalogy here.

Reply:
Yes, task == frame here, so it means 15 tasks and 15 frames. A chunk of 6 tasks is then also a chunk of 6 frames.